cloud-exec.md

```diff
@@ -0,0 +1,153 @@
+7. **Web run.** Execute an experiment from SaaS\Web using one of the methods
+   from above and DVC machinery.
+
+## Optimizations
+
+Multiple cloud instance optimization technicques might be used to optimize
+  the usage of resources or execution time:
+1. **Spot instances.**
+2. **Transparent spot instance.** Recover the execution if a spot instance
+   was terminated. DVC checkpoints and pipeline stages should be used for
+   preserving the state.
+3. **Volumes.** Volumes can be attached and reused in instances to minimize
+   data cache synchronization time.
+4. **Shared volumes.** Separate cloud services (such as Multi-Attach EBS or
+   EFS) might be needed for sharing data cache between multiple instances.
+
+## Monitoring 
+
+Additional features might be needed to improve user experienc:
+1. **State monitoring.** A user should see if an instance was already
+   allocated, what stage is running.
+2. **Metrics monitoring.**  Latest metrics monitoring (with some reasonable
+   delay - 10-30-60 seconds)
+
+
+## Resource orchestrators
+
+1. AWS
+2. Azure
+3. GCP (optional)
+4. Remote over SSH
+5. Kubernetes (K8S)
+6. HashiCorp Nomad (optional)
+7. Container services (ECS and Azure/GCP analogs)
+
+
+# Detailed design
+
+## Iterative Terraform provider
+
+The cloud management part seems like a heavy part of the project (from
+  the implementation point of view). However, the problem of executing and
+  managing cloud instances was solved in the CML project.
+
+CML separates out the user API part (commands) and the cloud instace
+  management code part.
+The cloud management is extracted to a separate **Iterative Terraform (TF)
+  provider**:
+  https://github.com/iterative/terraform-provider-iterative
+
+The Iterative provider is specific to data scientists, not DevOps (TF's
+  target audience). It has a bit simplified API and some ML specific
+  functionality (some of these are coming) like: recover spot-instance if
+  it was deleted or attach a reusable drive.
+
+The Iteratve TF provider can be used from DVC for cloud instace management.
+Some Python libraries might be required  for access to GoLang specific
+terraform-provider.
+
+Today the provider supports:
+- AWS, Azure, K8S orchestrators
+- Spot instances and Volumes optimizations.
+A few of the initial scenarios can be already implemented with these
+functionalities. More functionality is comming.
+
+## Instances configuration
+
+Instance configuration is a definition of an instance and its behavior:
+- orchestrator: AWS, GCP, Azure, remote, K8S, ...
+- instance type
+- volume
+- many other options
+
+Instance configuration is well developed in CML project and can be reused.
+
+## Executors and naming
+
+Many instance types and even clouds could be used in a single DVC project.
+We need a way of configuring **executors** and reusing them the same way
+  as user can configure a data remote.
+It can look like this:
+```bash
+$ dvc exp run --executor my-gpu-tesla
+...
```

Comment:
Would executors get setup like remotes?

```console
$ dvc executor add my-gpu-tesla ssh://...
$ dvc exp run --on my-gpu-tesla
```

Reply:
Are resulting outputs synchronized back into the workspace? Or does the executor keep it's own repo clone with separate state?

Reply:
> Are resulting outputs synchronized back into the workspace? Or does the executor keep it's own repo clone with separate state?

My assumption here would be that the results of an experiment are synced back into the workspace ("results" meaning git-tracked repo state). Current `--temp` exp runs are already (local) executors, and experiment execution is already structured so that it would work with any machine we can talk to via git+ssh. So the git-tracked results of an executor run will be retrieved as experiment refs (like with existing local `--temp` runs).

For DVC-tracked (cached) outputs, they would be fetched into the local cache. This could be done by either fetching the data directly from the executor machine, or by using an intermediate DVC remote (so after a run, the executor does `dvc push` to a DVC remote, and then the user's local machine does `dvc pull`). The final `dvc pull` could actually be optional here, since the user may not need the cache data on their local machine at all.

Using the intermediate remote seems like it would fit better to me, but I think that needs some more discussion.